### PR TITLE
Inactivity logout

### PR DESCRIPTION
**Pull request by calellowitz**
To log people out if they are inactive for 30 minutes as per the security audit
@TylerSheffels @sravfeyn cc: @snopoke @biyeun

### CODE REVIEW

File: corehq/middleware.py [hunk 1]
```diff
@@ -81,3 +85,14 @@ def process_response(self, request, response):
...
             profile_logger.info('{} time {}'.format(request.path, duration), extra={'duration': duration})
         return response
+
+
+class TimeoutMiddleware(object):
+
+    def process_request(self, request):
+        now = datetime.datetime.utcnow()
+        last_request = request.session.get('last_request')
+        if last_request is not None and now - string_to_utc_datetime(last_request) > datetime.timedelta(minutes=settings.INACTIVITY_TIMEOUT):
```
Comment (dimagimon) [inline corehq/middleware.py]: E501 line too long (141 > 115 characters)

**Comment by benrudolph** [comment]
oof this is kind of bummer, but suppose it's for the best. any chance we can have an "infinite" mode? this way we can overwrite the var in the `dev_settings` and have our local instance never log out

File: corehq/middleware.py (hunk 1 above)
Comment (snopoke) [inline corehq/middleware.py]: This if condition could be extracted to a function `if session_expired(request)`

File: settings.py [hunk 2]
```diff
@@ -147,6 +148,9 @@
 
 SESSION_ENGINE = "django.contrib.sessions.backends.cached_db"
 
+# time in minutes before forced logout due to inactivity
+INACTIVITY_TIMEOUT = 30
```
Comment (snopoke) [inline settings.py]: @benrudolph you can just make this `sys.maxint` for an infinite session

**Comment by snopoke** [comment]
Nothing blocking but expect followup PR

**Comment by czue** [comment]
@dimagi/product not sure whether you were already in the loop on this, but this now sets the HQ timeout to 30 minutes for everyone across the board when they are inactive (and then they are signed out).

@calellowitz was there ever any discussion about limiting that requirement to users who are members of specific domains? would be quite a bit more complex to implement, but might be nicer for a lot of our users who don't mind being permanently logged in.

also in terms of the security audit, we could just set a larger value on prod and a much smaller one on india if need be.

**Comment by benrudolph** [comment]
Thanks for the tip @snopoke. Cal I sill think it's worth updating the dev_settings to have that timeout to be unlimited. Invariably people are going to miss this PR and wonder why they're getting signed out all of sudden

**Comment by benrudolph** [comment]
just a heads up, `sys.maxint` doesn't work and renders this error: `normalized days too large to fit in a C int`

**Comment by czue** [comment]
https://github.com/dimagi/commcare-hq/pull/9306 is probably good enough

**Comment by czue** [comment]
@calellowitz @dimagi/product a number of people were very confused and unhappy about this behavior today so i updated the localsettings on prod to bump this up to 2 weeks. I think we should figure out reasonable comms or workarounds before dropping this down to 30 mins again.

**Comment by amsagoff** [comment]
Going to start a new thread off-PR

**Comment by amsagoff** [comment]
(let me know if you want to be added, wasn't sure who to include)

**Pull request merged**